text/3627-match-ergonomics-2024.md

```diff
@@ -0,0 +1,519 @@
+- Feature Name: `ref_pat_eat_one_layer_2024`
+- Start Date: 2024-05-06
+- RFC PR: [rust-lang/rfcs#3627](https://github.com/rust-lang/rfcs/pull/3627)
+- Rust Issue: [rust-lang/rust#123076](https://github.com/rust-lang/rust/issues/123076)
+
+# Summary
+[summary]: #summary
+
+Various changes to the match ergonomics rules:
+
+- On edition ≥ 2024, `&` and `&mut` patterns only remove a single layer of
+  references.
+- On edition ≥ 2024, `mut` on an identifier pattern does not force its binding
+  mode to by-value.
+- On all editions, `&` patterns can match against `&mut` references.
```

Comment:
This is a breaking change.

Rust uses the patterns for type inference, so changing what `&` can match will/should will stop some code from compiling. A simple example is

```rust
let &_ = generic_return(); // the function is inferred to return `&_`
```
([full working example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0e8440c88322611b90ed2f06fa999167))

Reply:
I hadn’t considered that case, but it actually *does* work under the current implementation: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=2acaaa1d97f85b4f2cb04f5160ee5848)

I’ll add tests and update the RFC, but for now I don’t think there is real breakage.

Reply:
Ahh, looking at the implementation I think I see what is happening. [This](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=1acce770a0ed215f7315679c5437016f) doesn’t work

Reply:
I’ve updated the RFC.

Reply:
> However, if the type of the scrutinee is unknown, an `&` pattern will still constrain inference to force it to be a shared reference. (97bf386e8e04e8267b706460602fb08046286f87)

I don't think this is right?

In the sense that while this is the currently implemented behavior, it's not a very good behavior. I think the inference part should be in tune/agreement with the matching part. If `&` can match `&mut`, it shouldn't guide inference towards using `&`.

Reply:
Match ergonomics interact poorly with inference everywhere: `let Foo(x) = function();` guides inference to get a `Foo` out of `function()` even when it could equally get a `&Foo`. I've often had spurious "type X is not Copy" errors because of a type error somewhere else that made inference ignore match ergonomics in this way.

Reply:
_Ideally_, we would constrain to `&` or `&mut` with priority for the former. But I don't think the compiler is smart enough for that